### ISSUE TITLE
Can i do training on combined dataset i.e. MPII and COCO and AI-challenger

### ISSUE THREAD

**Issue by eng100200** [leoxiaobin/deep-high-resolution-net.pytorch#125]
@leoxiaobin  My question is can i do training by combining all the available datasets, while, each of these datasets have different number of pose points? Or its not possible?

**Comment by eng100200** [comment]
@leoxiaobin hello, can you reply me? i  have also one more question from lib/dataset/mpii.py
![image](https://user-images.githubusercontent.com/28100251/67138841-db8f1f80-f27b-11e9-8915-38cdea99ba94.png)
why there is joints_3d mean？ mpii has 2d annotations or 3d?

**Comment by leoxiaobin** [comment]
For training hybrid datasets, you need to map different joints by yourself.

For MPII data set, we only use the 2d annotations.

**Comment by eng100200** [comment]
@leoxiaobin  How are you? thanks for your reply. I have not understand this piece of code, why we require shift and if this shift is performed we are losing the last heat map channel , i guess?
![image](https://user-images.githubusercontent.com/28100251/67193802-7af11580-f428-11e9-9f02-ecee290d4859.png)

**Comment by eng100200** [comment]
@leoxiaobin i have one more issue. Is there you have provided any code for generating ground truth heat maps for mpii validation set?

**Comment by eng100200** [comment]
@leoxiaobin hello, i want to share something with you and want to know  your comment on this. I have checked your JointsDataset. py code, the code generates the training samples after affine transform and wrapaffine to generate the final images for input to the model. Initially, i though that the center and scale is used to crop the original person from the image and creat a new image of of scale x scale size, and then later apply affine transforms on the newly formed image to augment the data...but, its different than i thought. It rather takes the original image itself and apply afine transform on the original image, and then forms a scale x scale dimension of image which is later fit into a 256 x 256 dimesion to form a 256 x 256 input training image. Below are the images generated from your JointsDataset. py alongwith original image is also posted:
Original image
![099992483](https://user-images.githubusercontent.com/28100251/68262118-4367a780-007d-11ea-89e1-fb1484b30840.jpg)
Generated two images for two labels:
![image_14015](https://user-images.githubusercontent.com/28100251/68262143-58443b00-007d-11ea-857b-a967f27d2873.jpg)
![image_14016](https://user-images.githubusercontent.com/28100251/68262161-61cda300-007d-11ea-80a7-f886ea81dfc6.jpg)
But, the two images generated i have two questions:
1. Why we do not crop each person and formed a new image and then later use all the affine transforms and resizing to 256 x 256 to form a image contain single person only?
2. If the training image contains two persons while there is only one label and the other person is considered background, does this not corrupt the training of model?; since, at one side we say a person, while, the other side we declare the other person as background.

**Comment by eng100200** [comment]
@ZP-Guo

**Comment by eng100200** [comment]
@ZP-Guo you can also look into this

**Comment by ZP-Guo** [comment]
The first question, maybe I can.
I guess you need to modify the number of joints from diffenrent datasets to the same number. and different dataset have different orders of joints so you need to modify it, too. And then choose a dataset as an example, i.e. MS COCO. Modify all formats to the MS COCO's and use the train and test code of MS COCO to do it.

**Comment by FightStone** [comment]
@leoxiaobin @ZP-Guo The best results in the paper were obtained by using other data. What other data were included? Looking forward to your reply!

**Comment by FightStone** [comment]
@eng100200 For your second question, i think if only one person is labeled according to the author's annotation, I think the model can learn the annotation rules of the author (for example, only learn the  person whose center point is in the middle of the picture).
Second, if all the keypoints in the picture are marked. So in the reference stage, how do you choose the target keypoint? Because each heatmap may have more than one peak!
The above is just my personal opinion，looking forward to official reply! @leoxiaobin

**Comment by eng100200** [comment]
@leoxiaobin @ZP-Guo One simple question is the MPII  is_visible labels are V=0 Not-Visible but labeled and V=1 Visible and labeled.

**Comment by annopackage** [comment]
@leoxiaobin @eng100200 
I am a little confused about the following code? 
-----
def get_affine_transform(
        center, scale, rot, output_size,
        shift=np.array([0, 0], dtype=np.float32), inv=0
):
    if not isinstance(scale, np.ndarray) and not isinstance(scale, list):
        print(scale)
        scale = np.array([scale, scale])

    scale_tmp = scale * 200.0
    src_w = scale_tmp[0] 
    dst_w = output_size[0]
    dst_h = output_size[1]

    rot_rad = np.pi * rot / 180
    src_dir = get_dir([0, src_w * -0.5], rot_rad)
    dst_dir = np.array([0, dst_w * -0.5], np.float32)

    src = np.zeros((3, 2), dtype=np.float32)
    dst = np.zeros((3, 2), dtype=np.float32)
    src[0, :] = center + scale_tmp * shift
    src[1, :] = center + src_dir + scale_tmp * shift
    dst[0, :] = [dst_w * 0.5, dst_h * 0.5]
    dst[1, :] = np.array([dst_w * 0.5, dst_h * 0.5]) + dst_dir

    src[2:, :] = get_3rd_point(src[0, :], src[1, :])
    dst[2:, :] = get_3rd_point(dst[0, :], dst[1, :])

    if inv:
        trans = cv2.getAffineTransform(np.float32(dst), np.float32(src))
    else:
        trans = cv2.getAffineTransform(np.float32(src), np.float32(dst))

    return trans
----
Why choose the three point coord to calculate the affine parameter?  Why they are paired points when doing affine transformation?
